File: projects/dejavu.cs/Dejavu.Language/CodeToken.cs
namespace Dejavu.Language;

/// <summary>Lexical token kinds inside <c>&lt;% ... %&gt;</c> (mirrors TS/Rust <c>CodeToken</c>).</summary>
public enum CodeTokenKind
{
    CodeEnd,
    PipeOp,
    OrOr,
    AndAnd,
    EqEq,
    NotEq,
    LessEq,
    GreaterEq,
    Less,
    Greater,
    Plus,
    Minus,
    Star,
    Slash,
    Percent,
    Bang,
    Dot,
    Comma,
    LParen,
    RParen,
    LBracket,
    RBracket,
    Bool,
    Null,
    In,
    Ident,
    Number,
    String,
}

/// <summary>A single code token with offsets relative to the code slice.</summary>
public readonly record struct CodeToken(
    CodeTokenKind Kind,
    string Text,
    bool? Value,
    int Start,
    int End);

File: projects/dejavu.kt/dejavu-language/src/main/kotlin/dejavu/language/CodeToken.kt
package dejavu.language

/** Lexical token kinds inside `<% ... %>` (mirrors TS / Rust `CodeToken`). */
enum class CodeTokenKind {
    CodeEnd,
    PipeOp,
    OrOr,
    AndAnd,
    EqEq,
    NotEq,
    LessEq,
    GreaterEq,
    Less,
    Greater,
    Plus,
    Minus,
    Star,
    Slash,
    Percent,
    Bang,
    Dot,
    Comma,
    LParen,
    RParen,
    LBracket,
    RBracket,
    Bool,
    Null,
    In,
    Ident,
    Number,
    String,
}

data class CodeToken(
    val kind: CodeTokenKind,
    /** Lexeme text (idents, numbers, string contents, bool literal text). */
    val text: String = "",
    val value: Boolean? = null,
    /** Offset within the code slice (not absolute source). */
    val start: Int,
    val end: Int,
)
